automates/model_analysis/identification_algorithms.py

```diff
@@ -0,0 +1,251 @@
+
+        # Line 6
+        if s_single in cc:
+            tree.call.line = 6
+            tree.call.s = s_single
+            s_single_length = len(s_single)
+            product_list = []
+            p_prod = gm.Probability()
+            for node in s_single:
+                node_topo_index = v.index(node)
+                cond_set = v[0:node_topo_index]
+                if p.product:
+                    p_prod = gm.parse_joint(p, node, cond_set, v, topo)
+                else:
+                    p_prod = deepcopy(p)
+                    p_prod.var = [node]
+                    p_prod.cond = cond_set
+                product_list.append(p_prod)
+            product_list.reverse()
+            if s_single_length > 1:
+                prob_new = gm.Probability(sumset=gm.ts(set(s_single) - set(y), topo), product=True, children= \
+                    product_list)
+                tree.root = prob_new
+                tree.call.id_check = True
+                return gm.ResultsInternal(p=prob_new, tree=tree)
+            if p_prod.product or p_prod.fraction:
+                p_prod.sumset = gm.ts(list(set(p_prod.sumset) | (set(s_single) - set(y))), topo)
+            else:
+                p_prod.var = gm.ts(set(p_prod.var) - (set(p_prod.sumset) | (set(s_single) - set(y))), topo)
+            tree.root = p_prod
+            tree.call.id_check = True
+            return gm.ResultsInternal(p=p_prod, tree=tree)
+
+        # Line 7
+        tree.call.s = s_single
+        s_prime = None
+        for component in cc:
+            if set(s_single).issubset(set(component)):
+                s_prime = component
+                break
+        tree.call.line = 7
+        tree.call.s_prime = s_prime
+        s_prime_length = len(s_prime)
+        g_s_prime = g.subgraph(s_prime)
+        g_s_prime_obs = gm.observed_graph(g_s_prime)
+        product_list = []
+        p_prod = gm.Probability()
+        for node in s_prime:
+            node_topo_index = v.index(node)
+            cond_set = v[0:node_topo_index]
+            if p.product:
+                p_prod = gm.parse_joint(p, node, cond_set, v, topo)
+            else:
+                p_prod = deepcopy(p)
+                p_prod.var = [node]
+                p_prod.cond = cond_set
+            product_list.append(p_prod)
+        product_list.reverse()
+        x_new = gm.ts(set(x) & set(s_prime), topo)
+        if s_prime_length > 1:
+            nxt = compute_ID(y, x_new, gm.Probability(product=True, children=product_list), g_s_prime, g_s_prime_obs, \
+                             s_prime, topo, gm.TreeNode())
+        else:
+            nxt = compute_ID(y, x_new, product_list[0], g_s_prime, g_s_prime_obs, s_prime, topo, gm.TreeNode())
+        tree.children.append(nxt.tree)
+        tree.call.id_check = nxt.tree.call.id_check
+        return gm.ResultsInternal(p=nxt.p, tree=tree)
+
+
+def compute_IDC(y, x, z, p, g, g_obs, v, topo, tree):
+    if len(p.var) == 0:
```

Comment:
@aswinchester see my above comment(s) about code duplication under conditionals and then update this instance and any other instances in this file.

Reply:
This has been completed